File: checkdigit.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "checkdigit.h"

int main(int argc, char const *argv[]) {
	if (argc != 2) {
		printf("You must provide one card number to check.\n");
		return 2;
	}
	if (strlen(argv[1]) != 16) {
		printf("Invalid credit card number.\n");
		return 2;
	}
	if (checkdigit(strtol(argv[1], NULL, 10))) {
		printf("The provided credit card number is valid.\n");
		return 0;
	} else {
		printf("The provided credit card number is invalid.\n");
		return 1;
	}
}

int checkdigit(long int num) {
	int sum = 0;
	int check = num % 10; // Get last digit, i.e. the checkdigit
	num /= 10; // Integer math truncates, so this just strips off the last digit

	for (int i = 1; i < 16; ++i) {
		int current = num % 10;

		// If the number is five or bigger, it times two will be 10 + some single digit number
		if (i % 2) {
			if (current >= 5) {
				sum += 1;
			}
			sum += ((2*current) % 10);
		} else {
			sum += current;
		}

		num /= 10;
	}
	
	// Find the number we need to add to the checksum to make it mod 10 = 0
	sum = (10 - (sum % 10)) % 10;

	return (sum == check ? 1 : 0);
}

File: checkdigit.js
// Please note that just because a card number is 'valid' it does not necessarily have to be an actual card.
// The checkdigit hashing function is quite vulnerable to collision attacks.

// For an explanation of how it works, see https://www.youtube.com/watch?v=hWkAe9FsfiE

const checkdigit = (number) => {
	checksum = 0;

	for(i = 0; i < 15; i++) {
		digit = parseInt(number.charAt(i));
		if(i % 2 == 0) {
			// If the number is five or bigger, it times two will be 10 + some single digit number
			if(digit >= 5){
				checksum += 1;
			}
			checksum += ((2*digit) % 10);
		} else {
			checksum += digit;
		}
	}

	// Find the number we need to add to the checksum to make it mod 10 = 0
	const calcuatedCheckDigit = (10 - (checksum % 10)) % 10;

	// Does the provided checkdigit (last digit of card) and what the checkdigit should be match?
	return (parseInt(number.charAt(15)) === calcuatedCheckDigit);
}

module.exports = { checkdigit }
